ipfs-8-to-9/migration/swapper.go

```diff
@@ -0,0 +1,165 @@
+package mg8
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	log "github.com/ipfs/fs-repo-migrations/stump"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	ds "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	dshelp "github.com/ipfs/go-ipfs-ds-help"
+	"github.com/multiformats/go-multihash"
+)
+
+// SyncSize specifies how much we batch data before committing and syncing
+// Increasing this number may result
+var SyncSize uint64 = 10 * 1024 * 1024 // 1MiB
+
+// NWorkers sets the number of batching threads to run
+var NWorkers int = 4
+
+// CidSwapper reads all the keys in a datastore and replaces
+// them with their raw multihash.
+type CidSwapper struct {
+	Store ds.Batching // the datastore to migrate.
+}
+
+// Run lists all the keys in the datastore and triggers a swap operation for
+// those corresponding to CIDv1s (replacing them by their raw multihash).
+//
+// Run returns the total number of keys swapped.
+func (cswap *CidSwapper) Run() (uint64, error) {
+	// Always perform a final sync
+	defer cswap.Store.Sync(ds.NewKey("/"))
+	// Query all keys. We will loop all keys
+	// and swap those that can be parsed as CIDv1.
+	queryAll := query.Query{
+		KeysOnly: true,
+	}
+
+	results, err := cswap.Store.Query(queryAll)
+	if err != nil {
+		return 0, err
+	}
+	defer results.Close()
+	resultsCh := results.Next()
+
+	var total uint64
+	var nErrors uint64
+	var wg sync.WaitGroup
+	wg.Add(NWorkers)
+	for i := 0; i < NWorkers; i++ {
+		go func() {
+			defer wg.Done()
+			n, e := cswap.swapWorker(resultsCh)
+			atomic.AddUint64(&total, n)
+			atomic.AddUint64(&nErrors, e)
+		}()
+	}
+	wg.Wait()
+	if nErrors > 0 {
+		return total, errors.New("errors happened during the migration. Consider running it again")
+	}
+
+	return total, nil
+}
+
+// swapWorkers reads query results from a channel and renames CIDv1 keys to
+// raw multihashes by reading the blocks and storing them with the new
+// key. Returns the number of keys swapped and the number of errors.
+func (cswap *CidSwapper) swapWorker(resultsCh <-chan query.Result) (uint64, uint64) {
+	var swapped uint64
+	var errored uint64
+	var curSyncSize uint64
+
+	// Process keys from the results channel
+	for res := range resultsCh {
+		if res.Error != nil {
+			log.Error(res.Error)
+			errored++
+			continue
+		}
+
+		oldKey := ds.NewKey(res.Key)
+		c, err := dsKeyToCid(oldKey)
+		if err != nil {
+			// complain if we find anything that is not a CID but
+			// leave it as it is.
+			log.Log("could not parse %s as a Cid", oldKey)
+			continue
+		}
+		if c.Version() == 0 { // CidV0 are multihashes, leave them.
+			continue
+		}
+
+		// Cid Version > 0
+		newKey := multihashToDsKey(c.Hash())
+		size, err := cswap.swap(oldKey, newKey)
+		if err != nil {
+			log.Error("swapping %s for %s: %s", oldKey, newKey, err)
+			errored++
+			continue
+		}
+		swapped++
+		curSyncSize += size
+
+		// Commit and Sync if we reached SyncSize
+		if curSyncSize >= SyncSize {
```

Comment:
Alternatively we could batch based on number of keys migrated. Total size seemed a better option though.

Reply:
I agree, total size seems like the way to go.

Reply:
Oh, wait. I'm not entirely sure if this is correct. The idea behind batching was to:

1. Add N new blocks in a batch.
2. Commit and Sync.
4. Remove the old versions in a batch
4. Commit and Sync.

I don't think the datastores (e.g., badger) guarantee ordering of events until a sync happens. Otherwise, we'd just sync once at the end and be done with it.

TL;DR: periodic syncing doesn't really buy us anything.

Reply:
> I don't think the datastores (e.g., badger) guarantee ordering of events until a sync happens

You mean you can be left in a state where the last write operation in a transaction succeeded but the previous ones were lost if something happens mid-sync? Well that's shitty..

The problem with `adding N blocks` to a batch and commit is that, if blocks are very small, we will be batching really small sizes. And we need to keep N low enough to not trigger `too many files opened` in flatfs, so it is going to be a low N given there are 4 workers.

How about we do two full passes, one adding all the new blocks, syncing by size as now, then another removing the old blocks?

Reply:
> You mean you can be left in a state where the last write operation in a transaction succeeded but the previous ones were lost if something happens mid-sync? Well that's shitty..

Not in a _transaction_. In a transaction, we'll be fine. However, batches aren't _technically_ transactions. They are in badger (implementation detail), but not in flatfs.

On the other hand, this logic is _probably_, technically safe in most/all existing case.

* Badger: Batches are actually transactions.
* Flatfs: When applying a batch, we put first, sync, then delete.
* Levelds: I think the batch is applied in-order then synced at the end.

> How about we do two full passes, one adding all the new blocks, syncing by size as now, then another removing the old blocks?

That would be safe. However, my concern there is that we'll double the size of the datastore (worst case).

Note: we don't actually need to use _batches_. We could also:

1. Add up to N megabytes.
2. Sync.
3. Remove the keys we just added.
4. Repeat.

Basically, your solution but we do it iteratively (without batches).

Reply:
So I added up to SyncSize, then Sync, then delete, then Sync, then repeat.

Reply:
The delete part is effectively a flatfs Batch, which makes me a bit scared about the file descriptor limit, but I haven't hit issues.

Reply:
Should be fine. Delete doesn't open a file descriptor (just sends the path itself down to the kernel).